verification/src/solidity/verifier.rs

```diff
@@ -311,10 +420,137 @@ impl Verifier {
+    }
+
+    /// Compares the result of local contract compilation with data specified on initialization.
+    ///
+    /// On success returns a tuple where first argument is a contract ABI, and the second
+    /// is constructor arguments passed on actual contract initialization.
+    fn compare(
+        &self,
+        contract: &Contract,
+    ) -> Result<(ethabi::Contract, Option<Bytes>), VerificationError> {
+        let deployed_bytecode = {
+            let bytes = contract
+                .get_deployed_bytecode_bytes()
+                .ok_or(VerificationError::MissedLibrary)?;
+            DeployedBytecode::try_from(bytes.0.clone())
+                .map_err(|err| VerificationError::InvalidDeployedBytecode(err.to_string()))?
+        };
+        let bytecode = {
+            let bytes = contract.get_bytecode_bytes().expect(
+                "contract compiled locally and has valid deployed bytecode: bytecode must be valid",
+            );
+            Bytecode::<CompilationResult>::try_from_bytes(bytes.0.clone(), &deployed_bytecode)
+                .expect(
+                "contract compiled locally and has valid deployed bytecode: bytecode must be valid",
+            )
+        };
+        let abi = contract
+            .get_abi()
+            .expect("contract was compiled locally: abi must exist");
+
+        self.check_metadata_hash_solc_versions(&deployed_bytecode)?;
+
+        self.bc_creation_tx_input
+            .verify_bytecode_with_extra_data(&bytecode)?;
+
+        let constructor_args = self.extract_constructor_args(abi.constructor(), &bytecode)?;
+
+        Ok((abi.into_owned(), constructor_args))
+    }
+
+    /// Checks that solc versions obtained from metadata hash correspond
+    /// for provided deployed bytecode and deployed bytecode obtained
+    /// as a result of local compilation.
+    fn check_metadata_hash_solc_versions(
+        &self,
+        deployed_bytecode: &DeployedBytecode,
+    ) -> Result<(), VerificationError> {
+        let compiled_solc = &deployed_bytecode.metadata_hash().solc;
+        let bc_solc = &self.bc_deployed_bytecode.metadata_hash().solc;
+        if bc_solc != compiled_solc {
+            let compiled_solc = compiled_solc
+                .as_ref()
+                .map(|b| DisplayBytes::from(b.clone()).to_string());
+            let bc_solc = bc_solc
+                .as_ref()
+                .map(|b| DisplayBytes::from(b.clone()).to_string());
+            return Err(VerificationError::CompilerVersionMismatch(Mismatch::new(
+                compiled_solc,
+                bc_solc,
+            )));
+        }
+        Ok(())
+    }
+
+    /// Extracts constructor arguments from the creation transaction input specified on
+    /// [`Verifier`] initialization.
+    ///
+    /// Returns `Err` if constructor arguments cannot be extracted (should not be the case
+    /// if `Bytecode.verify_bytecode_with_extra_data` was called before).
+    fn extract_constructor_args(
```

Comment:
why do you have `constructor_args` as `impl Bytecode`, and `extract_constructor_args` as `impl Verifier`?

Reply:
They are a little bit different. `constructor_args` is mostly just a sequence of bytes that happens to occur in `Bytecode` after all metadata and aux data information. Thus, we obtain that sequence of bytes from `Bytecode` structure. At the same time `extract_constructor_args` is more related to `Verifier` as it gets that sequence obtained from `Bytecode` and ensure that it is valid constructor args according to abi of the contract.